File: test/Coypu.Tests/When_interacting_with_the_browser/When_finding_then_clicking_any_element.cs
using System;
using System.Linq;
using Coypu.Tests.TestDoubles;
using NUnit.Framework;

package Coypu.Tests.When_interacting_with_the_browser
{
    [TestFixture]
    public class When_finding_then_clicking_any_element : BrowserInteractionTests
    {
        @Test
        public void It_makes_robust_call_to_find_then_clicks_element_on_underlying_driver()
        {
            var element = new StubElement();
            driver.StubCss("something.to click", element, browserSession);
            spyRobustWrapper.AlwaysReturnFromRobustly(element);

            var elementScope = browserSession.FindCss("something.to click");

            assertThat(driver.FindCssRequests, Is.Empty, "Finder not called robustly");

            elementScope.Click();

            RunQueryAndCheckTiming();

            assertThat(driver.FindCssRequests.Any(), Is.False, "Scope finder was not deferred");

            assertThat(driver.ClickedElements, Has.Member(element));
        }
    }
}

File: src/test/java/Coypu.Tests/When_interacting_with_the_browser/When_finding_then_clicking_any_element.java
package Coypu.Tests.When_interacting_with_the_browser;

import Coypu.Element;
import Coypu.Tests.TestDoubles.StubElement;
import org.junit.Test;

import static Coypu.Tests.When_interacting_with_the_browser.IsEmpty.empty;
import static org.hamcrest.Matchers.hasItem;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;

public class When_finding_then_clicking_any_element extends BrowserInteractionTests
{
    @Test
    public void it_makes_robust_call_to_find_then_clicks_element_on_underlying_driver()
    {
        Element element = new StubElement();
        driver.stubCss("something.to click", element, browserSession);
        spyRobustWrapper.alwaysReturnFromRobustly(element);

        elementScope = browserSession.findCss("something.to click");

        assertThat("Finder not called robustly", driver.FindCssRequests, is(empty()));

        elementScope.click();

        runQueryAndCheckTiming();

        assertThat("Scope finder was not deferred", driver.FindCssRequests, is(empty()));

        assertThat(driver.ClickedElements, hasItem(element));
    }
}
